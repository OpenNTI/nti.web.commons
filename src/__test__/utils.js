/*eslint-env jest*/
import renderer from 'react-test-renderer';

export const render = (exp, ...options) => renderer.create(exp, ...options);
export const verify = (exp, ...options) => {
	const r = render(exp, ...options);
	expect(r.toJSON()).toMatchSnapshot();
	return r;
};
