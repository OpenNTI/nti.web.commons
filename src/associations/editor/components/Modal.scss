@import "~@nti/style-common";

.associations-editor-modal-wrapper > .icon-close { display: none; }

.association-editor-modal {
	width: 100vw;
	max-width: 768px;
	margin: 0 auto;
	position: relative;
	background: white;

	.panel-title-bar {
		box-shadow: inset 0 -1px 0 0 var(--border-grey-light);
	}

	.content {
		height: calc(80vh - 20px);
		background: var(--panel-background);
		overflow: auto;
		padding: 1.25rem;

		@include opacity-on-covered-modal;
	}

	.dialog-buttons {
		background: white;
		box-shadow: inset 0 1px 0 0 var(--border-grey-light);

		&::before {
			display: none;
		}

		@include opacity-on-covered-modal;
	}
}
