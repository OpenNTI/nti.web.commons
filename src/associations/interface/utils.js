import Group from './Group';

export function groupDestinations (destinations) {
	let currentWrapper = null;

	return destinations.reduce((acc, destination) => {
		if (destination instanceof Group) {
			currentWrapper = null;
			acc.push(destination);
		} else {
			if (!currentWrapper) {
				currentWrapper = new Group();
				acc.push(currentWrapper);
			}

			currentWrapper.push(destination);
		}

		return acc;
	}, []);
}

export function flattenGroups (groups) {
	return groups.reduce((acc, group) => {
		if (group instanceof Group) {
			acc = [...acc, ...flattenGroups(group.items)];
		} else {
			acc.push(group);
		}

		return acc;
	}, []);
}


export function filterGroups (groups, fn) {
	return groups.reduce((acc, group) => {
		const filteredGroup = group.filter(fn);

		if (!filteredGroup.isEmpty) {
			acc.push(filteredGroup);
		}

		return acc;
	}, []);
}


export function groupItemsByParent (items) {
	function getLabel (item) {
		return item.label || item.title;
	}

	const parents = items.reduce((acc, item) => {
		const parent = item.item.parent();
		const id = parent.getID();

		if (acc.map[id]) {
			acc.map[id].push(item);
		} else {
			acc.order.push(id);
			acc.map[id] = new Group (getLabel(parent), [item]);
		}

		return acc;
	}, {map: {}, order: []});

	return parents.order.map(x => parents.map[x]);
}
