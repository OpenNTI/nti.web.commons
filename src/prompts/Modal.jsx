import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import isIOS from '@nti/util-ios-version';
import { declareCustomElement } from '@nti/lib-dom';

import LockScroll from '../components/LockScroll';

declareCustomElement('dialog');


const needsSafariFix = e => e && (isIOS() && /input|textarea|select/i.test(e.target.tagName));
const stopEvent = e => e && e.stopPropagation();

/**
 * This is the Modal Shell of a dialog or prompt. Do not use this directly unless
 * you really want to manage the lower levels.
 *
 * If you are looking to lauch a modal, see {@link Dialog}
 *
 * @private This class should be considered impementation detail of ModalManager.
 * @class Modal
 */
export default class Modal extends React.Component {

	static propTypes = {
		/** @ignore */
		children: PropTypes.any,
		/** @ignore */
		className: PropTypes.string,

		/**
		 * A ModalManager instance.
		 * @private
		 * @ignore
		 * @type {ModalManager}
		 */
		Manager: PropTypes.object.isRequired,

		/**
		 * Sent by ModalManager
		 * @ignore
		 * @type {Boolean}
		 */
		isPortal: PropTypes.bool,

		/**
		 * A callback to call on requesting a dismissal.
		 *
		 * @private
		 * @ignore
		 * @type {function}
		 */
		onDismiss: PropTypes.func.isRequired,

		/**
		 * Enable the click listener on the mask that will call onDismiss.
		 *
		 * @private
		 * @ignore
		 * @type {boolean}
		 */
		closeOnMaskClick: PropTypes.bool,

		/**
		 * Don't center the dialog. Position it at the top instead.
		 *
		 * @private
		 * @ignore
		 * @type {boolean}
		 */
		tall: PropTypes.bool
	}


	static childContextTypes = {
		close: PropTypes.func
	}

	/* @ignore */
	attachContentRef = x => this.content = x


	getChildContext = () => ({
		close: this.close
	})


	constructor (props) {
		super(props);

		this.props.Manager.addUpdateListener(this.onManagerUpdate);
		this.state = { staging: true };
	}

	componentDidMount () {
		// Staging is here because the ref content isn't set when the modal manager calls isHidden in the render.
		// We set state for another render to occur and correctly call isHidden with a set ref.
		this.setState({ staging: false }, () => this.focus());
	}

	componentWillUnmount () {
		this.props.Manager.removeUpdateListener(this.onManagerUpdate);
	}


	onManagerUpdate = () => !this.props.isPortal && this.forceUpdate()


	/**
	 * @param {object} e - the event
	 * @returns {undefined}
	 */
	close = (e) => {
		const {onDismiss} = this.props;

		stopEvent(e);

		/* istanbul ignore else */
		if (onDismiss) {
			onDismiss(e);
		}
	}


	focus = () => {
		this.content.focus();
	}


	/**
	 * @param {object} e - the event
	 * @returns {undefined}
	 */
	onMaskClick = (e) => {
		const {closeOnMaskClick} = this.props;
		if (closeOnMaskClick) {
			this.close(e);
		}
	}

	/**
	 * @param {object} e - the event
	 * @returns {undefined}
	 */
	onFocus = (e) => {
		if (needsSafariFix(e)) {
			this.setState({safariFix: true});
		}
	}

	/**
	 * @param {object} e - the event
	 * @returns {undefined}
	 */
	onBlur = (e) => {
		if (needsSafariFix(e)) {
			this.setState({safariFix: false});
		}
	}


	render () {
		const {Manager, children, className, tall} = this.props;
		const {safariFix} = this.state;
		const hidden = Manager.isHidden(this);

		const classes = cx('modal-mask', className, {hidden, 'safari-fix': safariFix, tall});

		const ios = isIOS();
		const timeout = 500;


		return (
			<>
				<LockScroll />
				<TransitionGroup>
					<CSSTransition
						key={className}
						classNames="modal-mask"
						timeout={timeout}
						appear={!ios}
						enter={!ios}
						exit={!ios}
					>
						<div
							className={classes}
							onFocus={this.onFocus}
							onBlur={this.onBlur}
							onClick={this.onMaskClick}
						>
							<i className="icon-close" onClick={this.close}/>
							<dialog role="dialog" className="modal-content" open ref={this.attachContentRef} tabIndex="-1" onClick={stopEvent}>
								{React.Children.map(children, child =>
									typeof child.type === 'string'
										? child
										: React.cloneElement(child, { onDismiss: this.close })
								)}
							</dialog>
						</div>
					</CSSTransition>
				</TransitionGroup>
			</>
		);
	}

}
