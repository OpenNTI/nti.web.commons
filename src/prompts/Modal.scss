@import "~@nti/style-common";

.modal-content {
	max-height: 100vh;
}

.modal-mask {
	@include touch-scroll;
	overflow: {
		x: hidden;
		y: auto;
	}

	position: fixed;
	top: 0;
	left: 0;
	bottom: 0;
	right: 0;
	background: rgba(0, 0, 0, 0.7);
	perspective: 300px;
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;

	body.inline-dialog-open & {
		z-index: 1;
	}

	body > .modal + .modal & {
		background: none;
	}

	&.tall {
		justify-content: flex-start;
	}

	.modal-content {
		animation-delay: 0.15s;
		flex: 0 0 auto;

		& > * {
			margin: 0 auto;
			position: relative;

			@include covered-modal {
				&::after {
					content: '';
					position: absolute;
					top: 0;
					bottom: 0;
					left: 0;
					right: 0;
					background: rgba(255, 255, 255, 0.5);
				}
			}
		}
	}

	&.safari-fix {
		position: absolute !important;
		bottom: auto;

		.modal-content {
			animation: none;
		}
	}

	.icon-close {
		position: absolute;
		width: 16px;
		height: 16px;
		background-image: url(./assets/close.png);
		background-repeat: no-repeat;
		background-size: cover;
		right: 1rem;
		top: 1rem;
		cursor: pointer;
	}

	dialog {
		display: block;
		position: static;
		left: auto;
		right: auto;
		width: auto;
		height: auto;
		color: inherit;
		margin: initial;
		border-width: initial;
		border-style: initial;
		border-color: initial;
		border-image: initial;
		padding: initial;
		background: none;
	}
}

.modal-mask-appear,
.modal-mask-enter {
	opacity: 0.01;

	html:not(.is-ie) & {
		transition: opacity 300ms ease-out; //make this transition 10ms longer to let the .modal-content animation playout
	}
}

.modal-mask-appear.modal-mask-appear-active,
.modal-mask-enter.modal-mask-enter-active {
	opacity: 1;

	.modal-content {
		animation: modal-content 250ms ease both;
	}
}

.modal-mask-exit {
	opacity: 1;
	transition: opacity 250ms ease-out;
}

.modal-mask-exit.modal-mask-exit-active {
	opacity: 0.01;
}

@keyframes modal-content {
	0% {
		transform: translate3d(0, 0, -30px);
		opacity: 0;
	}

	100% {
		transform: translate3d(0, 0, 0);
		opacity: 1;
	}
}
