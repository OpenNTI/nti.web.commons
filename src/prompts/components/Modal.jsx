import React from 'react';
import cx from 'classnames';
import Transition from 'react-addons-css-transition-group';
import isIOS from 'nti-util-ios-version';
import {declareCustomElement} from 'nti-lib-dom';

import LockScroll from '../../components/LockScroll';

import Manager from '../ModalManager';

declareCustomElement('dialog');

const stop = e => e.stopPropagation();

const needsSafariFix = e => e && (isIOS() && /input|textarea|select/i.test(e.target.tagName));


export default class Modal extends React.Component {

	static propTypes = {
		onDismiss: React.PropTypes.func.isRequired,
		children: React.PropTypes.node,
		className: React.PropTypes.string,
		closeOnMaskClick: React.PropTypes.bool,
		tall: React.PropTypes.bool
	}


	static childContextTypes = {
		close: React.PropTypes.func
	}


	getChildContext = () => ({
		close: this.close
	})


	state = {}


	componentDidMount () {
		Manager.addUpdateListener(this.onManagerUpdate);
		this.focus();
	}


	componentWillUnmount () {
		Manager.removeUpdateListener(this.onManagerUpdate);
	}


	onManagerUpdate = () => this.forceUpdate()


	close = (e) => {
		const {onDismiss} = this.props;
		if (e) {
			e.stopPropagation();
		}
		if (onDismiss) {
			onDismiss(e);
		}
	}


	focus = () => {
		this.content.focus();
	}


	onMaskClick = (e) => {
		const {closeOnMaskClick} = this.props;
		if (closeOnMaskClick && (e.target === this.mask || e.target === this.content)) {
			this.close(e);
		}
	}


	onFocus = (e) => {
		if (needsSafariFix(e)) {
			this.setState({safariFix: true});
		}
	}


	onBlur = (e) => {
		if (needsSafariFix(e)) {
			this.setState({safariFix: false});
		}
	}


	render () {
		const {children, className, tall} = this.props;
		const {safariFix} = this.state;
		const hidden = Manager.isHidden(this);

		const classes = cx('modal-mask', className, {hidden, 'safari-fix': safariFix, tall});

		return (
			<Transition transitionName="modal-mask"
				transitionAppear
				transitionLeave
				transitionAppearTimeout={500}
				transitionEnterTimeout={500}
				transitionLeaveTimeout={500}
			>
				<LockScroll />
				<div ref={(x) => this.mask = x} className={classes}
					onFocus={this.onFocus}
					onBlur={this.onBlur}
					onTouchStart={stop}
					onTouchMove={stop}
					onClick={this.onMaskClick}
					>
					<i className="icon-close" onClick={this.close}/>
					<dialog role="dialog" className="modal-content" open ref={x => this.content = x} tabIndex="-1">
						{React.cloneElement(
							React.Children.only(children),
							{ onDismiss: this.close }
						)}
					</dialog>
				</div>
			</Transition>
		);
	}

}
