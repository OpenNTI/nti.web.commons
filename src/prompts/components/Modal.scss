@import "~nti-style-common";

.modal-mask-appear,
.modal-mask-enter {
	opacity: 0.01;
	transition: opacity .26s ease-out; //make this transition 10ms longer to let the .modal-content animation playout
}

.modal-mask-appear.modal-mask-appear-active,
.modal-mask-enter.modal-mask-enter-active {
	opacity: 1;
	.modal-content {
		animation: modal-content .25s ease both;
	}
}

.modal-mask-leave {
	opacity: 1;
	transition: opacity .25s ease-out;
}

.modal-mask-leave.modal-mask-leave-active {
	opacity: 0.01;
}


.modal-mask {
	@include touch-scroll;
	overflow: { x: hidden; y: scroll; }
	position: fixed;
	top: 0;
	left: 0;
	bottom: 0;
	right: 0;
	background: rgba(0,0,0,.7);
	perspective: 300px;

	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;

	body.inline-dialog-open & {
		z-index: 1;
	}

	&.tall {
		justify-content: flex-start;
	}

	&.safari-fix {
		position: absolute !important;
		.modal-content {
			animation: none;
		}
	}

	.icon-close {
		position: absolute;
		width: 16px;
		height: 16px;
		background-image: url(./assets/close.png);
		background-repeat: no-repeat;
		background-size: cover;
		right: 1rem;
		top: 1rem;
		cursor: pointer;
	}
}

.modal-mask dialog {
	display: block;
	position: static;
	left: auto;
	right: auto;
	width: auto;
	height: auto;
	color: inherit;
	margin: initial;
	border-width: initial;
	border-style: initial;
	border-color: initial;
	border-image: initial;
	padding: initial;
	background: none;
}

.modal-content {
	animation-delay: .15s;
	flex: 0 0 auto;

	& > * {
		margin: 0 auto;

		@include covered-modal {
			&::before {
				content: '';
				position: absolute;
				top: 0; bottom: 0;
				left: 0; right: 0;
				background: rgba(255, 255, 255, 0.5);
			}
		}
	}

}

@keyframes modal-content {
	0% {
		transform: translate3d(0, 0, -30px);
		opacity: 0;
	}

	100% {
		transform: translate3d(0,0,0);
		opacity: 1;
	}
}


body > .modal + .modal {
	.modal-mask {
		background: none;
	}
}
