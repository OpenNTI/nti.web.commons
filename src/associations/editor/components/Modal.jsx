import React from 'react';
import PropTypes from 'prop-types';

import {modal} from '../../../prompts';
import TitleBar from '../../../components/panels/TitleBar';
import DialogButtons from '../../../components/DialogButtons';

import Editor from './View';

AssociationEditorModal.propTypes = {
	title: PropTypes.string,
	onDismiss: PropTypes.func
};
export default function AssociationEditorModal (props) {
	const {title, onDismiss, ...otherProps} = props;
	const buttons = [
		{label: 'Done', onClick: onDismiss}
	];

	return (
		<div className="association-editor-modal">
			<TitleBar title={title} iconAction={onDismiss} />
			<div className="content">
				<Editor {...otherProps} />
			</div>
			<DialogButtons buttons={buttons} />
		</div>
	);
}

export function openEditorModal (title, associations, filterFn, getString, beforeClose = () => {}) {
	modal(
		(
			<AssociationEditorModal title={title} associations={associations} filterFn={filterFn} getString={getString} />
		),
		{
			className: 'associations-editor-modal-wrapper',
			onBeforeDismiss () {
				beforeClose();
			}
		}
	);
}
