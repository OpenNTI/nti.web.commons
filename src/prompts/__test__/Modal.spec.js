/* eslint-env jest */
jest.mock('react-dom', () => require('../../__mocks__/react-dom.disabled'));
jest.mock('@nti/util-ios-version', () => () => global.simulateIOS);
jest.mock('@nti/lib-dom', () => Object.assign(
	require.requireActual('@nti/lib-dom'), {
		getScrollPosition: () => ({
			top: 100,
			left: 0,
		})
	}));

import React from 'react';

import {verify} from '../../__test__/utils';
import Modal from '../Modal';

const createNodeMock = ({type}) => document.createElement(type);


describe('Modals', () => {
	const mockManager = () => ({
		addUpdateListener: jest.fn(),
		removeUpdateListener: jest.fn(),
		isHidden: jest.fn(() => false)
	});

	afterEach(() => {
		delete global.simulateIOS;
	});

	test ('Simple Mount & Unmount', () => {
		const manager = mockManager();
		const dismiss = jest.fn();
		verify(<Modal Manager={manager} onDismiss={dismiss}/>, {createNodeMock}).unmount();

		expect(dismiss).not.toHaveBeenCalled();
		expect(manager.isHidden).toHaveBeenCalled();

		expect(manager.addUpdateListener).toHaveBeenCalled();

		const {mock: {calls: [args]}} = manager.addUpdateListener;

		expect(manager.removeUpdateListener).toHaveBeenCalledWith(...args);
	});

	test ('Close Event', () => {
		const manager = mockManager();
		const dismiss = jest.fn();
		const renderer = verify(<Modal Manager={manager} onDismiss={dismiss}/>, {createNodeMock});

		const cmp = renderer.getInstance();

		expect(() => cmp.close()).not.toThrow();
		expect(dismiss).toHaveBeenCalledWith(undefined);

		const event = {
			stopPropagation () {}
		};
		cmp.close(event);
		expect(dismiss).toHaveBeenCalledWith(event);

		renderer.unmount();
	});

	test ('Clicks inside modal do not trigger mask clicks', () => {
		const manager = mockManager();
		const dismiss = jest.fn();
		const renderer = verify(<Modal Manager={manager} onDismiss={dismiss}/>, {createNodeMock});

		const cmp = renderer.getInstance();

		const event = {
			stopPropagation: jest.fn()
		};

		jest.spyOn(cmp, 'onMaskClick');

		renderer.root.findByType('dialog').props.onClick(event);

		expect(cmp.onMaskClick).not.toHaveBeenCalled();
		expect(event.stopPropagation).toHaveBeenCalled();

		renderer.unmount();
	});

	test ('Mask clicks may close', () => {
		const manager = mockManager();
		const dismiss = jest.fn();
		const expression = <Modal Manager={manager} onDismiss={dismiss} />;
		const renderer = verify(expression, {createNodeMock});
		const cmp = renderer.getInstance();
		jest.spyOn(cmp, 'onMaskClick');

		//We need to re-render to get the spyed onMaskClick function
		renderer.update(React.cloneElement(expression));


		const event = {
			stopPropagation: jest.fn()
		};


		renderer.root.find(x => x.type === 'div' && x.props.onClick && /mask/.test(x.props.className)).props.onClick(event);

		expect(cmp.onMaskClick).toHaveBeenCalled();
		expect(event.stopPropagation).not.toHaveBeenCalled();

		renderer.update(React.cloneElement(expression, {closeOnMaskClick: true}));

		cmp.onMaskClick.mockClear();
		event.stopPropagation.mockClear();

		renderer.root.find(x => x.type === 'div' && x.props.onClick && /mask/.test(x.props.className)).props.onClick(event);

		expect(cmp.onMaskClick).toHaveBeenCalled();
		expect(event.stopPropagation).toHaveBeenCalled();

		renderer.unmount();
	});

	test ('Safari Hack', () => {
		const manager = mockManager();
		const dismiss = jest.fn();
		const renderer = verify(<Modal Manager={manager} onDismiss={dismiss}/>, {createNodeMock});

		const cmp = renderer.getInstance();

		cmp.onFocus({target: {tagName: 'input'}});
		expect(cmp.state.safariFix).toBeUndefined();
		cmp.onBlur({target: {tagName: 'input'}});
		expect(cmp.state.safariFix).toBeUndefined();

		global.simulateIOS = true;
		cmp.onFocus({target: {tagName: 'input'}});
		expect(cmp.state.safariFix).toBe(true);
		cmp.onBlur({target: {tagName: 'input'}});
		expect(cmp.state.safariFix).toBe(false);

		renderer.unmount();
	});

});
