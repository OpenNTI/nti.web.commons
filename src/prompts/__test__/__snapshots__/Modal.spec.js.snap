// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modals Clicks inside modal do not trigger mask clicks 1`] = `
<div>
  <div
    className="modal-mask"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
  >
    <i
      className="icon-close"
      onClick={[Function]}
    />
    <dialog
      className="modal-content"
      onClick={[Function]}
      open={true}
      role="dialog"
      tabIndex="-1"
    />
  </div>
</div>
`;

exports[`Modals Close Event 1`] = `
<div>
  <div
    className="modal-mask"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
  >
    <i
      className="icon-close"
      onClick={[Function]}
    />
    <dialog
      className="modal-content"
      onClick={[Function]}
      open={true}
      role="dialog"
      tabIndex="-1"
    />
  </div>
</div>
`;

exports[`Modals Mask clicks may close 1`] = `
<div>
  <div
    className="modal-mask"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
  >
    <i
      className="icon-close"
      onClick={[Function]}
    />
    <dialog
      className="modal-content"
      onClick={[Function]}
      open={true}
      role="dialog"
      tabIndex="-1"
    />
  </div>
</div>
`;

exports[`Modals Safari Hack 1`] = `
<div>
  <div
    className="modal-mask"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
  >
    <i
      className="icon-close"
      onClick={[Function]}
    />
    <dialog
      className="modal-content"
      onClick={[Function]}
      open={true}
      role="dialog"
      tabIndex="-1"
    />
  </div>
</div>
`;

exports[`Modals Simple Mount & Unmount 1`] = `
<div>
  <div
    className="modal-mask"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
  >
    <i
      className="icon-close"
      onClick={[Function]}
    />
    <dialog
      className="modal-content"
      onClick={[Function]}
      open={true}
      role="dialog"
      tabIndex="-1"
    />
  </div>
</div>
`;
